lib/rate.js

```diff
@@ -0,0 +1,204 @@
+ * This makes Rate useful for looping at a regular frequency (hz). Rate#sleep()
+ * avoids blocking the JS event-loop by returning a Promise that the caller
+ * should block on, e.g. use 'await rate.sleep()'.
+ *
+ * Note that Rate.sleep() does not prevent rclnodejs from invoking callbacks
+ * such as a subscription or client if the entity's node is spinning. Thus
+ * if your intent is to use rate to synchronize when callbacks are invoked
+ * then use a spinOnce() just after rate.sleep() as in the example below.
+ *
+ * Rate runs within it's own private rcl context. This enables it to be
+ * available immediately after construction. That is, Rate does not require a
+ * spin or spinOnce to be active.
+ *
+ * @example
+ * async function run() {
+ *   await rclnodejs.init();
+ *   const node = rclnodejs.createNode('mynode');
+ *   const rate = node.createRate(1); // 1 hz
+ *   while (true) {
+ *     doSomeStuff();
+ *     await rate.sleep();
+ *     rclnodejs.spinOnce(node);
+ *   }
+ * }
+ */
+class Rate {
+  /**
+   * Create a new instance.
+   * @hideconstructor
+   * @param {number} hz - The frequency (hz) between (0.0,1000] hz,
+   *   default = 1 hz
+   */
+  constructor(hz = 1) {
+    if (typeof hz !== 'number') {
+      throw new TypeError('Invalid argument');
+    }
+
+    const MAX_RATE_HZ_IN_MILLISECOND = 1000.0;
+    if (hz <= 0.0 || hz > MAX_RATE_HZ_IN_MILLISECOND) {
+      throw new RangeError(
+        `Hz must be between 0.0 and ${MAX_RATE_HZ_IN_MILLISECOND}`
+      );
+    }
+
+    this._hz = hz;
+    const period = Math.round(1000 / hz);
+    this._rateTimer = new RateTimer(period);
+  }
+
+  /**
+   * Get the frequency in hertz (hz) of this timer.
+   *
+   * @returns {number} - hertz
+   */
+  get frequency() {
+    return this._hz;
+  }
+
+  /**
+   * Returns a Promise that when waited on, will block the sender
+   * until the end of the current timer cycle.
+   *
+   * If the Rate has been cancelled, calling this method will
+   * result in an error.
+   *
+   * @example
+   * (async () => {
+   *   await rate.sleep();
+   * })();
+   *
+   * @returns {Promise} - Wait on the promise will delay the sender
+   * (not the Node event-loop) until the end of the current timer cycle.
+   */
+  async sleep() {
+    if (this.isCanceled()) {
+      throw new Error('Rate has been cancelled.');
+    }
+
+    return new Promise(resolve => {
+      this._rateTimer.sleep(() => {
+        resolve();
+      });
+    });
+  }
+
+  /**
+   * Permanently stops the rate's timer.
+   *
+   * @returns {undefined}
+   */
+  cancel() {
+    this._rateTimer.cancel();
+  }
+
+  /**
+   * Determine if this rate has been cancelled.
+   *
+   * @returns {boolean} - True when cancel() has been called; False otherwise.
+   */
+  isCanceled() {
+    return this._rateTimer.isCanceled();
+  }
+}
+
+const NOP_FN = () => {};
+
+let NODE_CNT = 1;
+function genNodeName() {
+  return 'rate_context' + NODE_CNT++;
```

Comment:
Should the name be prefixed with the name of the node that created the Rate? Otherwise, it looks like if you ran 2 separate JS scripts that create a Rate the names would conflict since NODE_CNT wouldn't be shared.

Reply:
Yes, absolutely! Good suggestion.